docs/dev_guide/code_coverage_guide.md

```diff
@@ -0,0 +1,104 @@
+
+#### Internal
+`alpini`, `venice-avro-compatibility-test`, `venice-client-common`, `venice-common`, `venice-consumer`, 
+`venice-jdk-compatibility-test`, `venice-test-common`
+
+#### Services
+
+`venice-router`, `venice-controller`, `venice-server`
+
+## Submodule-level coverage verification
+
+### Description
+This will check the overall code coverage at the submodule level.
+Jacoco generates the report based off the current submodule, perform the coverage verification and fail the build if the
+coverage is below the threshold. The threshold, targets at the branch coverage, is defined in the `build.gradle` of each 
+submodule independently.
+
+### Example commands
+```Bash
+# Template
+./gradlew :<module name>:<submodule name>:jacocoTestCoverageVerification 
+
+./gradlew :clients:venice-push-job:jacocoTestCoverageVerification 
+```
+
+The report will be located at `<module name>/<submodule name>/build/reports/jacoco/test/index.html`.
+
+
+### Debugging notes
+
+#### Run commands against module doesn't generate the report
+
+Due to the current project setup, running the commands, e.g. `./gradlew :clients:jacocoTestCoverageVerification`, doesn't
+execute the unit tests thus no jacoco report will be generated. Please be sure to run the commands against the submodule.
+
+## New-commit coverage verification
+
+### Description
+This will check the overall code coverage at the commit-level.
+DiffCoverage, which is an extension of Jacoco, gathers the diff files by comparing the local branch and remote upstream 
+main branch, and leverage the Jacoco report, to re-generate a new report only for these newly added lines/files. Similarly,
+it performs the coverage verification and fail the build if the coverage is below the threshold. The threshold is defined 
+at 60% for branch coverage.
+
+### Example commands
+```Bash
+# Template
+./gradlew :<module name>:<submodule name>:jacocoTestCoverageVerification diffCoverage
+
+./gradlew :clients:venice-push-job:jacocoTestCoverageVerification diffCoverage 
+```
+
+The report will be located at `<module name>/build/reports/jacoco/diffCoverage/html/index.html`.
+
+### Debugging notes
+
+#### Integ tests are added but DiffCoverage doesn't identify the code coverage of it
+Though Integ tests is strongly encouraged, both Jacoco and DiffCoverage only work with unit tests so please consider writing 
+unit tests.
+
+#### Unit tests are added but the DiffCoverage report doesn't reflect my changes
+
+There are two possible reasons.
+1. **Jacoco report isn't up-to-date**. DiffCoverage report relies on the Jacoco Report to reflect the correct coverage. 
+If the Jacoco report isn't up-to-date, for example, executing the wrong command `./gradlew :clients:venice-push-job:diffCoverage` which
+misses the step of re-running unit tests and updating the Jacoco report, can cause this issue. 
+   1. Please be sure to run test and Jacoco report first.
```

Comment:
Can we make `diffCoverage` depend on `jacocoTestReport`?

Reply:
I think we can and I was trying to do it in #168 but my limited Gradle knowledge couldn't help me figure out some task dependencies so after spending too much time on this, I think it's acceptable to make it not depend on `jacocoTestReport` 😅  If you have any insights, please lmk!

Reply:
I think I see the problem. `diffCoverage` needs to run on a per-module basis, whereas `jacocoTestReport` runs on the submodule level.

If that's the case, it is okay to leave it as it is for now